File: src/main/java/com/github/frierenzk/utils/ShellUtils.java
package com.github.frierenzk.utils;

import java.io.*;
import java.util.Iterator;
import java.util.List;

public class ShellUtils {
    private BufferedReader inputBuffer, errorBuffer;
    private Process process;

    public BufferedReader getInputBuffer() {
        return inputBuffer;
    }

    public BufferedReader getErrorBuffer() {
        return errorBuffer;
    }

    public void exec(List<String> command) {
        var processBuilder = new ProcessBuilder(command);
        try {
            process = processBuilder.start();
        } catch (Exception exception) {
            System.out.println(exception.toString());
            inputBuffer = new BufferedReader(new StringReader(exception.toString()));
            return;
        }
        inputBuffer = new BufferedReader(new InputStreamReader(process.getInputStream()));
        errorBuffer = new BufferedReader(new InputStreamReader(process.getErrorStream()));
    }
    
    public void execCommands(List<String> commands) throws IOException {
        var processBuilder = new ProcessBuilder("sh");
        process = processBuilder.start();
        inputBuffer = new BufferedReader(new InputStreamReader(process.getInputStream()));
        errorBuffer = new BufferedReader(new InputStreamReader(process.getErrorStream()));
        var stdin = new BufferedWriter(new OutputStreamWriter(process.getOutputStream()));
        for (var i : commands) {
            stdin.write(i);
            stdin.newLine();
            stdin.flush();
        }
        stdin.close();
    }

    public int getReturnCode() throws InterruptedException {
        if (process != null) {
            if (process.isAlive()) {
                return process.waitFor() + 1;
            }
            return process.exitValue() + 1;
        }
        return -1;
    }

    public boolean getAlive() {
        if (process != null) return process.isAlive();
        else return false;
    }

    public int terminate() {
        if (process == null) return -1;
        process.destroy();
        try {
            process.wait(3000);
        } catch (Exception ignored) {
        }
        if (process.isAlive()) process.destroyForcibly();
        try {
            process.wait();
        } catch (Exception ignored) {
        }
        return process.exitValue();
    }
}

File: src/main/kotlin/com/github/frierenzk/utils/ShellUtils.kt
package com.github.frierenzk.utils

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.io.BufferedReader
import java.io.BufferedWriter
import java.nio.file.attribute.PosixFilePermission
import java.nio.file.attribute.PosixFilePermissions
import java.util.concurrent.TimeUnit
import kotlin.io.path.bufferedWriter
import kotlin.io.path.createTempFile
import kotlin.io.path.deleteIfExists
import kotlin.io.path.pathString

object ShellUtils {
    fun exec(command: List<String>): Process = object : Process() {
        override val process by lazy {
            try {
                ProcessBuilder("bash", "-c", command.joinToString(" ")).start()
            } catch (exception: Exception) {
                exception.printStackTrace()
                null
            }
        }

        override var outBuffer: BufferedReader? = process?.inputStream?.bufferedReader()
        override var errorBuffer: BufferedReader? = process?.errorStream?.bufferedReader()
        override var writer: BufferedWriter? = process?.outputWriter()
    }

    @JvmName("execCommands")
    fun exec(commands: List<List<String>>): Process = object : Process() {
        override val process by lazy {
            try {
                val file = createTempFile(suffix = ".sh",
                    attributes = arrayOf(PosixFilePermissions.asFileAttribute(setOf(
                        PosixFilePermission.OWNER_WRITE,
                        PosixFilePermission.OWNER_READ,
                        PosixFilePermission.OWNER_EXECUTE))))
                file.bufferedWriter().use { writer ->
                    commands.map { it.joinToString(" ") }.forEach {
                        writer.appendLine(it)
                    }
                    writer.flush()
                }
                ProcessBuilder("bash", "-c", "\"${file.pathString}\"").start().also {
                    CoroutineScope(Dispatchers.IO).launch {
                        delay(3 * 1000L)
                        runCatching { file.deleteIfExists() }.onFailure { it.printStackTrace() }
                    }
                }
            } catch (exception: Exception) {
                exception.printStackTrace()
                null
            }
        }

        override var outBuffer: BufferedReader? = process?.inputStream?.bufferedReader()
        override var errorBuffer: BufferedReader? = process?.errorStream?.bufferedReader()
        override var writer: BufferedWriter? = process?.outputWriter()
    }

    abstract class Process {
        protected abstract val process: java.lang.Process?
        abstract var outBuffer: BufferedReader?
        abstract var errorBuffer: BufferedReader?
        abstract var writer: BufferedWriter?
        val returnCode: Int
            get() = try {
                process?.waitFor(3, TimeUnit.SECONDS)
                process?.exitValue() ?: -1
            } catch (exception: IllegalThreadStateException) {
                exception.printStackTrace()
                -1
            }

        fun stop(): Int {
            closeHandle(process?.toHandle())
            if (process?.isAlive == true)
                process?.waitFor(3000, TimeUnit.MILLISECONDS)
            return try {
                process?.exitValue() ?: -1
            } catch (ignore: IllegalThreadStateException) {
                -1
            }
        }

        private fun closeHandle(p: ProcessHandle?) {
            p?.descendants()?.forEach { closeHandle(it) }
            if (p?.supportsNormalTermination() == true) p.destroy()
            else p?.destroyForcibly()
        }

        val isAlive: Boolean
            get() = process?.isAlive ?: false
    }
}